File: java/hidingplaces.java
import java.util.Scanner;

public class hidingplaces {
	static int[][] chessBoard;
	static int level, maxlevel;

	static void singleFill(int row, int col, int level) {
		chessBoard[row][col] = level;
		if (col >= 2) {
			if (row != 0)
				fillToList(row - 1, col - 2, level);
			if (row != 7)
				fillToList(row + 1, col - 2, level);
		}
		if (row <= 5) {
			if (col != 0)
				fillToList(row + 2, col - 1, level);
			if (col != 7)
				fillToList(row + 2, col + 1, level);
		}
		if (col <= 5) {
			if (row != 0)
				fillToList(row - 1, col + 2, level);
			if (row != 7)
				fillToList(row + 1, col + 2, level);
		}
		if (row >= 2) {
			if (col != 0)
				fillToList(row - 2, col - 1, level);
			if (col != 7)
				fillToList(row - 2, col + 1, level);
		}
	}

	static void fillToList(int row, int col, int level) {
		if (chessBoard[row][col] == 0 || chessBoard[row][col] > level)
			singleFill(row, col, ++level);
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();
		while (n-- > 0) {
			chessBoard = new int[8][8];
			level = maxlevel = 1;
			String knight = scan.next();
			int col = knight.charAt(0) - 'a';
			int row = 8 - Character.getNumericValue(knight.charAt(1));
			singleFill(row, col, level);
			for (int i = 0; i < 8; i++)
				for (int j = 0; j < 8; j++)
					maxlevel = Math.max(maxlevel, chessBoard[i][j]);
			System.out.print((maxlevel - 1) + " ");
			for (int i = 0; i < 8; i++)
				for (int j = 0; j < 8; j++)
					if (chessBoard[i][j] == maxlevel)
						System.out.print(String.valueOf((char) ('a' + j)) + (8 - i) + " ");
			System.out.println();
		}
		scan.close();
	}
}

File: ts/src/hidingplaces.ts
import readline = require('readline')

class ChessBoard {
  private SIZE = 8
  private cells: number[][]

  constructor () {
    this.cells = Array(this.SIZE)
      .fill([])
      .map(() => Array(this.SIZE).fill(0))
  }

  public static toPos (pos: string): [number, number] {
    return [8 - parseInt(pos.charAt(1)), pos.charCodeAt(0) - 'a'.charCodeAt(0)]
  }

  public static fromPos (x: number, y: number): string {
    return `${String.fromCharCode(y + 'a'.charCodeAt(0))}${8 - x}`
  }

  public get maxValue (): number {
    return this.cells.reduce((max, row) => Math.max(max, ...row), 0)
  }

  public get maxCells (): [number, number][] {
    const result: [number, number][] = []
    for (let x = 0; x < this.SIZE; x++) {
      for (let y = 0; y < this.SIZE; y++) {
        if (this.get(x, y) === this.maxValue) {
          result.push([x, y])
        }
      }
    }
    return result
  }

  public fill (x: number, y: number, value = 1) {
    this.set(x, y, value)

    const next: [number, number][] = []
    if (y >= 2) {
      if (x !== 0) {
        next.push([x - 1, y - 2])
      }
      if (x !== 7) {
        next.push([x + 1, y - 2])
      }
    }
    if (x <= 5) {
      if (y !== 0) {
        next.push([x + 2, y - 1])
      }
      if (y !== 7) {
        next.push([x + 2, y + 1])
      }
    }
    if (y <= 5) {
      if (x !== 0) {
        next.push([x - 1, y + 2])
      }
      if (x !== 7) {
        next.push([x + 1, y + 2])
      }
    }
    if (x >= 2) {
      if (y !== 0) {
        next.push([x - 2, y - 1])
      }
      if (y !== 7) {
        next.push([x - 2, y + 1])
      }
    }

    next.forEach(([x, y]) => this.fillNext(x, y, value))
  }

  private fillNext (x: number, y: number, value = 1) {
    if (this.get(x, y) === 0 || this.get(x, y) - value > 1) {
      this.fill(x, y, value + 1)
    }
  }

  private get (x: number, y: number): number {
    return (this.cells[x] ?? [])[y] ?? 0
  }

  private set (x: number, y: number, value: number): void {
    (this.cells[x] ?? [])[y] = value
  }
}

function generateAnswers (): Map<string, string> {
  const answers: Map<string, string> = new Map()
  for (let char = 'a'.charCodeAt(0); char <= 'h'.charCodeAt(0); char++) {
    for (let num = 1; num <= 8; num++) {
      const key = `${String.fromCharCode(char)}${num}`

      const board = new ChessBoard()
      const [x, y] = ChessBoard.toPos(key)
      board.fill(x, y)

      const value = `${board.maxValue - 1} ${board.maxCells.reduce(
        (v, c) => v.concat(`${ChessBoard.fromPos(...c)} `),
        ''
      )}`

      answers.set(key, value)
    }
  }
  return answers
}

const answers = generateAnswers()
let lineCount = 0

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount > 0) {
      console.log(answers.get(line) ?? 'NA')
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
